src/WebJobs.Extensions.Http/HttpDirectRequestBindingProvider.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Net.Http;
+using System.Reflection;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Microsoft.Azure.WebJobs.Host.Bindings;
+using Microsoft.Azure.WebJobs.Host.Config;
+using Microsoft.Azure.WebJobs.Host.Protocols;
+using Newtonsoft.Json;
+using static Microsoft.Azure.WebJobs.Extensions.Http.HttpTriggerAttributeBindingProvider.HttpTriggerBinding;
+
+namespace Microsoft.Azure.WebJobs.Extensions.Http
+{
+    // support allowing binding any non-attributed HttpRequestMessage parameter to the incoming http request.     
+    //    Foo([HttpTrigger] Poco x, HttpRequestMessage y); 
+    internal class HttpDirectRequestBindingProvider : IBindingProvider
+    {
+        public Task<IBinding> TryCreateAsync(BindingProviderContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException("context");
+            }
+
+            ParameterInfo parameter = context.Parameter;
+            if (parameter.ParameterType == typeof(HttpRequestMessage))
+            {
+                // Not already claimed by another trigger?
+                if (parameter.GetCustomAttributes(false).Length == 0)
+                {
+                    return Task.FromResult<IBinding>(new HttpRequestBinding());
+                }
+            }
+            return Task.FromResult<IBinding>(null);
+        }
+
+        public class HttpRequestBinding : IBinding
+        {
+            public bool FromAttribute
+            {
+                get
+                {
+                    return false;
+                }
+            }
+
+            public Task<IValueProvider> BindAsync(BindingContext context)
+            {
+                if (context == null)
+                {
+                    throw new ArgumentNullException(nameof(context));
+                }
+                var request = context.BindingData["$request"];
```

Comment:
Didn't we have constants with special binding data names somewhere? Would be good to move this there.